### PR TITLE
Correction in Methods and Spellings

### PR DESCRIPTION
**Pull request by atishamte**
Some corrections in,
- Method return type
- Namespace
- Return Type
- Spellings
- Logics

**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [x] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Controller.php [hunk 1]
```diff
@@ -179,14 +179,14 @@ protected function loadHelpers()
...
-	 * @param array $messages An array of custom error messages
+	 * @param array|string  $rules
+	 * @param array         $messages An array of custom error messages
 	 *
 	 * @return boolean
 	 */
 	public function validate($rules, array $messages = []): bool
 	{
-		$this->validator = Services::validation();
+		$this->validator = \Config\Services::validation();
```
Comment (lonnieezell) [inline system/Controller.php]: Would prefer this was added at the top of the as a use statement. Not sure why it wasn't already.

File: system/Controller.php [hunk 2]
```diff
@@ -211,9 +211,8 @@ public function validate($rules, array $messages = []): bool
 		}
 
 		$success = $this->validator
-			->withRequest($this->request)
 			->setRules($rules, $messages)
-			->run();
+			->run($this->request->getPost());
```
Comment (lonnieezell) [inline system/Controller.php]: Were you running into an issue with the withRequest() method? If not, no need to change it.

File: system/Entity.php [hunk 3]
```diff
@@ -566,7 +566,7 @@ private function castAsJson($value, bool $asArray = false)
 
 				if (json_last_error() !== JSON_ERROR_NONE)
 				{
-					throw CastException::forInvalidJsonFormatException(json_last_error());
+					throw \CodeIgniter\Exceptions\CastException::forInvalidJsonFormatException(json_last_error());
```
Comment (lonnieezell) [inline system/Entity.php]: Add as a use statement at the top of the file, please.

File: system/HTTP/IncomingRequest.php [hunk 4]
```diff
@@ -152,14 +152,19 @@ class IncomingRequest extends Request
 	 * @param string                      $body
 	 * @param \CodeIgniter\HTTP\UserAgent $userAgent
 	 */
-	public function __construct($config, $uri = null, $body = 'php://input', UserAgent $userAgent)
+	public function __construct($config, $uri = null, $body = 'php://input', UserAgent $userAgent = null)
```
Comment (lonnieezell) [inline system/HTTP/IncomingRequest.php]: prefer to require UserAgent in this request. It it already added in the service method, so revert this change.

File: system/Validation/FormatRules.php [hunk 5]
```diff
@@ -196,7 +196,7 @@ public function numeric(string $str = null): bool
 	 *
 	 * @return boolean
 	 */
-	public function regex_match(string $str = null, string $pattern, array $data): bool
+	public function regex_match(string $str, string $pattern, array $data): bool
```
Comment (lonnieezell) [inline system/Validation/FormatRules.php]: As with all of the other validation rules, this should accept a null value.

File: system/Controller.php (hunk 2 above)
Comment (atishamte) [reply system/Controller.php]: the run method accepts the $data as a parameter. It was showing the syntactical error if run() is written. Hence changed.

File: system/HTTP/IncomingRequest.php (hunk 4 above)
Comment (atishamte) [reply system/HTTP/IncomingRequest.php]: Changed it because there is a concept, always keep optional parameter the end of parameter listing in function arguments. Hence changed.

File: system/Controller.php (hunk 2 above)
Comment (atishamte) [reply system/Controller.php]: This is an interface function prototype
public function run(array $data, string $group = null): bool;
where $data is not optional

File: system/HTTP/IncomingRequest.php (hunk 4 above)
Comment (lonnieezell) [reply system/HTTP/IncomingRequest.php]: I'm familiar with the concept. In this case, it doesn't work, though. Due to the different types of rules, and some wanting/needing more information than others, the first item should always be the $str. And sometimes that value will be null.

File: system/Controller.php (hunk 2 above)
Comment (lonnieezell) [reply system/Controller.php]: Then the correct solution here is to make $data nullable in the Validator's run method. And adjust tests accordingly. 

it's always possible for PUT or other HTTP verbs than POST to provide data to the application.

Comment (atishamte) [reply system/Controller.php]: As per the documentation, I have made the changes. It is written to validate only POST params.
Please refer the attachment and link

https://codeigniter4.github.io/userguide/incoming/controllers.html#validating-post-data

controllers codeigniter4 4 0 0 alpha 5 documentation

Do you want me to make changes for other HTTP verbs?

Comment (lonnieezell) [reply system/Controller.php]: Yes, that should be updated. The `withRequest` method handles any of the verbs, which is why that was used in the controller. Looks like I forgot to update the docs. 

Also - if you could break this out into it's own PR that would be great. Simpler to deal with things when it's only one change/type of change in a PR.

Comment (atishamte) [reply system/Controller.php]: Sure. I will make a separate PR for the same, where run method $data param is nullable.
I will update the doc as well.

**Comment by atishamte** [comment]
When It will get merged? I am unable to generate new PR

**Comment by jim-parry** [comment]
Patience, young padawan!
You should be creating PR features in separate branches, rather than in your develop branch. Then you would be able to have multiple changes in progress at the same time.

**Comment by atishamte** [comment]
Mistakenly I pushed it to develop, that's why I am asking.
Thanks

**Comment by atishamte** [comment]
All the changes have been done as per @lonnieezell suggested
